transformer_engine/common/transformer_engine.cpp

```diff
@@ -273,6 +273,284 @@ void CheckOutputTensor(const Tensor &t, const std::string &name, bool allow_empt
+             " first_dims has invalid shape");
+  NVTE_CHECK(!t.second_dims.shape.empty() && t.second_dims.shape[0] > 0, "Grouped tensor ", name,
+             " second_dims has invalid shape");
+  NVTE_CHECK(t.first_dims.shape[0] == t.second_dims.shape[0], "Grouped tensor ", name,
+             " first_dims and second_dims must have the same size (got ", t.first_dims.shape[0],
+             " vs ", t.second_dims.shape[0], ")");
+  NVTE_CHECK(t.first_dims.shape[0] == t.num_tensors(), "Grouped tensor ", name,
+             " shape arrays size (", t.first_dims.shape[0], ") doesn't match num_tensors (",
+             t.num_tensors(), ")");
+
+  // Check cumulative_tensor_sizes if it's set (optional but must be consistent)
+  if (t.has_cumulative_tensor_sizes()) {
+    // Ensure cumulative_tensor_sizes is 1D
+    NVTE_CHECK(t.cumulative_tensor_sizes.shape.size() == 1, "Grouped tensor ", name,
+               " cumulative_tensor_sizes must be 1D (got ", t.cumulative_tensor_sizes.shape.size(),
+               "D)");
+    // Ensure cumulative_tensor_sizes has correct dtype (size_t = Int64)
+    NVTE_CHECK(t.cumulative_tensor_sizes.dtype == DType::kInt64, "Grouped tensor ", name,
+               " cumulative_tensor_sizes must have dtype Int64 (got ",
+               to_string(t.cumulative_tensor_sizes.dtype), ")");
+    NVTE_CHECK(!t.cumulative_tensor_sizes.shape.empty() && t.cumulative_tensor_sizes.shape[0] > 1,
+               "Grouped tensor ", name, " cumulative_tensor_sizes has invalid shape");
+    NVTE_CHECK(t.cumulative_tensor_sizes.shape[0] == t.first_dims.shape[0] + 1, "Grouped tensor ",
+               name, " cumulative_tensor_sizes size (", t.cumulative_tensor_sizes.shape[0],
+               ") must be first_dims size + 1 (", t.first_dims.shape[0] + 1, ")");
+  }
+
+  // Additional validation for scale_inv based on scaling mode
+  if (is_tensor_scaling(t.scaling_mode)) {
+    NVTE_CHECK(t.scale.num_elements == expected_num_tensors, "Grouped tensor ", name,
+               " scale.num_elements (", t.scale.num_elements,
+               ") doesn't match expected num_tensors (", expected_num_tensors, ")");
+    NVTE_CHECK(t.amax.num_elements == expected_num_tensors, "Grouped tensor ", name,
+               " amax.num_elements (", t.amax.num_elements,
+               ") doesn't match expected num_tensors (", expected_num_tensors, ")");
+  } else if (t.scaling_mode == NVTE_NVFP4_1D_SCALING || t.scaling_mode == NVTE_MXFP8_1D_SCALING) {
+    if (t.has_data()) {
+      NVTE_CHECK(
+          !t.scaling_mode == NVTE_NVFP4_1D_SCALING || t.amax.num_elements == expected_num_tensors,
+          "Grouped tensor ", name, " amax.num_elements (", t.amax.num_elements,
+          ") doesn't match expected num_tensors (", expected_num_tensors, ")");
+      auto scale_shape = get_block_scale_shape(t.scaling_mode, t.data.sum_first_dims,
+                                               t.data.sum_second_dims, false);
```

Comment:
This is not true - you can't just pad the sum of the dimensions. The proper size is $128*4*\sum_i \lceil \frac{m_i}{128}\rceil\lceil\frac{k_i}{128}\rceil$ (note that I said size and not shape since the shape does not really have meaning here) which is more than that. In practice we do not really have the means (and frankly, I don't think we need that) to properly check that the allocations here are correct and should rely on the caller, who has additional information (like the fact that some dimensions are constant or that they just downloaded the dimensions to the CPU and checked), to do the right thing.